### PR TITLE
{AKS} Update addon preparation - part 1

### PR DESCRIPTION
**Pull request by FumingZhang**
**Description**

- Not long ago, in aks-preview @tilnl-ms helped sorted out the internal operations required by some addons ([PR #3911](https://github.com/Azure/azure-cli-extensions/pull/3911)), and we are gradually synchronizing this part.
- This would also fix a bug in the preparation process of monitoring addon.
- This first PR adds some basic functions, mainly around creating a role assignment.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/acs/_resourcegroup.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# --------------------------------------------------------------------------------------------
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_resourcegroup.py:1]: This file is basically the same as the [_resourcegroup.py](https://github.com/Azure/azure-cli-extensions/blob/main/src/aks-preview/azext_aks_preview/_resourcegroup.py) file in aks-preview.

The differences are as follows:
- sort imports
- change the error type

For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L4154) is the original implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
+# --------------------------------------------------------------------------------------------
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:1]: This file is basically the same as the [_roleassignments.py](https://github.com/Azure/azure-cli-extensions/blob/main/src/aks-preview/azext_aks_preview/_roleassignments.py) file in aks-preview,  except that the error type is changed to `AzCLIError` and add backward compatibility for creating role assignment.

File: src/azure-cli/azure/cli/command_modules/acs/_resourcegroup.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from azure.cli.core.azclierror import AzCLIError
+
+from ._client_factory import cf_resource_groups
+
+
+def get_rg_location(ctx, resource_group_name, subscription_id=None):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_resourcegroup.py:11]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L4154) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+from knack.log import get_logger
+from msrestazure.azure_exceptions import CloudError
+
+from ._client_factory import (get_auth_management_client,
+                              get_graph_rbac_management_client)
+
+logger = get_logger(__name__)
+
+
+def _get_object_stubs(graph_client, assignees):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:22]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1552) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    params = GetObjectsParameters(include_directory_object_references=True,
+                                  object_ids=assignees)
+    return list(graph_client.objects.get_objects_by_object_ids(params))
+
+
+def resolve_object_id(cli_ctx, assignee):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:28]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1532) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+    # 2+ matches should never happen, so we only check 'no match' here
+    if not result:
+        raise AzCLIError(
+            "No matches in graph database for '{}'".format(assignee))
+
+    return result[0].object_id
+
+
+def resolve_role_id(role, scope, definitions_client):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:50]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1512) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            raise AzCLIError("Role '{}' doesn't exist.".format(role))
+        if len(role_defs) > 1:
+            ids = [r.id for r in role_defs]
+            err = "More than one role matches the given name '{}'. Please pick a value from '{}'"
+            raise AzCLIError(err.format(role, ids))
+        role_id = role_defs[0].id
+    return role_id
+
+
+def build_role_scope(resource_group_name: str, scope: str, subscription_id: str):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:70]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1499) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+
+def _create_role_assignment(cmd, role, assignee,
+                            resource_group_name=None, scope=None, resolve_assignee=True):
+    factory = get_auth_management_client(cmd.cli_ctx, scope)
+    assignments_client = factory.role_assignments
+    definitions_client = factory.role_definitions
+
+    # FIXME: is this necessary?
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:108]: Is this necessary?

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        parameters = RoleAssignmentCreateParameters(
+            role_definition_id=role_id, principal_id=object_id)
+        return assignments_client.create(scope, assignment_name, parameters, custom_headers=custom_headers)
+
+    # for backward compatibility
+    RoleAssignmentProperties = get_sdk(cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION,
+                                       'RoleAssignmentProperties', mod='models',
+                                       operation_group='role_assignments')
+    properties = RoleAssignmentProperties(role_definition_id=role_id, principal_id=object_id)
+    return assignments_client.create(scope, assignment_name, properties, custom_headers=custom_headers)
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:140]: Added for old api versions.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            err = 'Resource group "{}" is redundant because scope is supplied'
+            raise AzCLIError(err.format(resource_group_name))
+    elif resource_group_name:
+        scope = subscription_scope + '/resourceGroups/' + resource_group_name
+    else:
+        scope = subscription_scope
+    return scope
+
+
+def create_role_assignment(
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:83]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1458) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        cmd,
+        role,
+        assignee,
+        resource_group_name,
+        scope,
+        resolve_assignee=(is_service_principal and resolve_assignee),
+    )
+
+
+def _create_role_assignment(cmd, role, assignee,
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:102]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1464) is the corresponding implementation in `custom.py`.

File: src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+    # for backward compatibility
+    RoleAssignmentProperties = get_sdk(cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION,
+                                       'RoleAssignmentProperties', mod='models',
+                                       operation_group='role_assignments')
+    properties = RoleAssignmentProperties(role_definition_id=role_id, principal_id=object_id)
+    return assignments_client.create(scope, assignment_name, properties, custom_headers=custom_headers)
+
+
+def add_role_assignment(cmd, role, service_principal_msi_id, is_service_principal=True, delay=2, scope=None):
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py:143]: For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L542) is the corresponding implementation in `custom.py`.

**Comment by yonzhan** [comment]
AKS

**Review by tilnl-ms — APPROVED** [review]

**Review by andyliuliming — APPROVED** [review]

**Review by wangzelin007 — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**